File: frontend/src/types/exam.ts
// Exam System Types
// Centralized type definitions for the exam system

// ============================================================================
// ENUMS AND CONSTANTS
// ============================================================================

export enum ExamStatus {
  DRAFT = 'draft',
  ACTIVE = 'active',
  ARCHIVED = 'archived',
}

export enum ExamType {
  // General Categories
  APTITUDE = 'aptitude',
  SKILL = 'skill',
  KNOWLEDGE = 'knowledge',
  PERSONALITY = 'personality',
  CUSTOM = 'custom',

  // Japanese Aptitude Tests (適性検査)
  SPI = 'spi', // Synthetic Personality Inventory
  TAMATEBAKO = 'tamatebako', // 玉手箱
  CAB = 'cab', // Computer Aptitude Battery
  GAB = 'gab', // General Aptitude Battery
  TG_WEB = 'tg_web', // TG-WEB
  CUBIC = 'cubic', // CUBIC
  NTT = 'ntt_aptitude', // NTT適性検査
  KRAEPELIN = 'kraepelin', // クレペリン検査
  SJT = 'sjt', // Situational Judgment Test

  // Technical/Industry-Specific
  PROGRAMMING_APTITUDE = 'programming_aptitude', // プログラミング適性検査
  NUMERICAL_ABILITY = 'numerical_ability', // 数理能力検査
  VERBAL_ABILITY = 'verbal_ability', // 言語能力検査
  LOGICAL_THINKING = 'logical_thinking', // 論理思考テスト
}

export enum QuestionType {
  MULTIPLE_CHOICE = 'multiple_choice',
  SINGLE_CHOICE = 'single_choice',
  TEXT_INPUT = 'text_input',
  ESSAY = 'essay',
  TRUE_FALSE = 'true_false',
  RATING = 'rating',
  MATCHING = 'matching',
}

export enum SessionStatus {
  NOT_STARTED = 'not_started',
  IN_PROGRESS = 'in_progress',
  COMPLETED = 'completed',
  EXPIRED = 'expired',
  SUSPENDED = 'suspended',
}

// ============================================================================
// CORE EXAM INTERFACES
// ============================================================================

export interface Exam {
  id: number;
  title: string;
  description: string | null;
  exam_type: string;
  status: string;
  company_id: number | null;  // Nullable for global exams
  created_by: number | null;
  time_limit_minutes: number | null;
  max_attempts: number;
  passing_score: number | null;
  is_randomized: boolean;
  is_public: boolean;  // Public exam flag
  allow_web_usage: boolean;
  monitor_web_usage: boolean;
  require_face_verification: boolean;
  face_check_interval_minutes: number;
  show_results_immediately: boolean;
  show_correct_answers: boolean;
  show_score: boolean;
  instructions: string | null;
  created_at: string;
  updated_at: string;
  // Computed fields
  total_questions?: number | null;
  total_sessions?: number | null;
  completed_sessions?: number | null;
  average_score?: number | null;
}

export interface ExamInfo {
  id: number;
  title: string;
  description: string | null;
  exam_type: string;
  status: string;
  company_id: number | null;
  created_by: number | null;
  time_limit_minutes: number | null;
  max_attempts: number;
  passing_score: number | null;
  is_randomized: boolean;
  is_public: boolean;
  allow_web_usage: boolean;
  monitor_web_usage: boolean;
  require_face_verification: boolean;
  face_check_interval_minutes: number;
  show_results_immediately: boolean;
  show_correct_answers: boolean;
  show_score: boolean;
  instructions: string | null;
  created_at: string;
  updated_at: string;
  total_questions: number;
}

// ============================================================================
// QUESTION INTERFACES
// ============================================================================

export interface Question {
  id: number;
  question_text: string;
  question_type: string;
  order_index: number;
  points: number;
  time_limit_seconds: number | null;
  is_required: boolean;
  options: Record<string, string> | null;
  correct_answers: string[] | null;
  max_length: number | null;
  min_length: number | null;
  rating_scale: number | null;
  explanation: string | null;
  tags: string[] | null;
}

export interface QuestionFormData {
  question_text: string;
  question_type: string;
  points: number;
  time_limit_seconds: number | null;
  is_required: boolean;
  options: Record<string, string>;
  correct_answers: string[];
  max_length: number | null;
  min_length: number | null;
  rating_scale: number | null;
  explanation: string;
  tags: string[];
}

// ============================================================================
// SESSION INTERFACES
// ============================================================================

export interface SessionInfo {
  id: number;
  exam_id: number;
  candidate_id: number;
  assignment_id: number | null;
  status: string;
  attempt_number: number;
  started_at: string | null;
  completed_at: string | null;
  expires_at: string | null;
  time_remaining_seconds: number | null;
  current_question_index: number;
  total_questions: number;
  questions_answered: number;
  score: number | null;
  max_score: number | null;
  percentage: number | null;
  passed: boolean | null;
  web_usage_detected: boolean;
  web_usage_count: number;
  face_verification_failed: boolean;
  face_check_count: number;
  exam_title: string;
  exam_type: string;
  require_face_verification: boolean;
  face_check_interval_minutes: number;
  monitor_web_usage: boolean;
  allow_web_usage: boolean;
  // Additional fields for results
  candidate_name?: string;
  candidate_email?: string;
  time_spent_minutes?: number | null;
}

export interface SessionSummary {
  id: number;
  candidate_name: string;
  candidate_email: string;
  attempt_number: number;
  status: string;
  started_at: string | null;
  completed_at: string | null;
  score: number | null;
  percentage: number | null;
  passed: boolean | null;
  web_usage_detected: boolean;
  face_verification_failed: boolean;
  time_spent_minutes: number | null;
}

// ============================================================================
// ANSWER INTERFACES
// ============================================================================

export interface Answer {
  question_id: number;
  answer_data?: Record<string, any>;
  answer_text?: string;
  selected_options?: string[];
  time_spent_seconds?: number;
}

export interface AnswerInfo {
  id: number;
  session_id: number;
  question_id: number;
  answer_data: Record<string, any> | null;
  answer_text: string | null;
  selected_options: string[] | null;
  is_correct: boolean | null;
  points_earned: number;
  points_possible: number;
  time_spent_seconds: number | null;
  answered_at: string;
}

// ============================================================================
// ASSIGNMENT INTERFACES
// ============================================================================

export interface ExamAssignment {
  id: number;
  exam_id: number;
  candidate_id: number;
  assigned_by: number | null;
  due_date: string | null;
  custom_time_limit_minutes: number | null;
  custom_max_attempts: number | null;
  custom_is_randomized: boolean | null;
  is_active: boolean;
  completed: boolean;
  notification_sent: boolean;
  reminder_sent: boolean;
  created_at: string;
  updated_at: string;
  // Related info
  exam_title: string;
  exam_type: string;
  sessions_count: number;
  latest_session: {
    id: number;
    status: string;
    percentage: number | null;
    passed: boolean | null;
    started_at: string | null;
    completed_at: string | null;
  } | null;
}

// ============================================================================
// MONITORING INTERFACES
// ============================================================================

export interface MonitoringEvent {
  id: number;
  session_id: number;
  event_type: string;
  event_data: any;
  severity: string;
  timestamp: string;
}

// ============================================================================
// FORM DATA INTERFACES
// ============================================================================

export interface ExamFormData {
  title: string;
  description: string;
  exam_type: string;
  company_id?: number | null;  // Optional for system admin (null = global exam)
  is_public: boolean;  // Public exam flag
  is_global?: boolean;  // Global exam flag (system admins only)
  time_limit_minutes: number | null;
  max_attempts: number;
  passing_score: number | null;
  is_randomized: boolean;
  allow_web_usage: boolean;
  monitor_web_usage: boolean;
  require_face_verification: boolean;
  face_check_interval_minutes: number;
  show_results_immediately: boolean;
  show_correct_answers: boolean;
  show_score: boolean;
  instructions: string;
}

// ============================================================================
// API RESPONSE INTERFACES
// ============================================================================

export interface ExamListResponse {
  exams: Exam[];
  total: number;
  page: number;
  page_size: number;
  has_more: boolean;
}

export interface ExamTakeResponse {
  session: SessionInfo;
  questions: Question[];
  current_question: Question | null;
  time_remaining_seconds: number | null;
  can_navigate: boolean;
}

export interface ExamResults {
  session: SessionInfo;
  answers: AnswerInfo[];
  questions?: Question[];
  monitoring_events?: MonitoringEvent[];
}

export interface ExamStatistics {
  exam_id: number;
  total_assigned: number;
  total_started: number;
  total_completed: number;
  completion_rate: number;
  average_score: number | null;
  average_time_minutes: number | null;
  pass_rate: number | null;
  question_statistics: Array<{
    question_number?: number;
    question_type?: string;
    correct_percentage?: number;
    average_time_seconds?: number;
    [key: string]: any;
  }>;
}

// ============================================================================
// COMPONENT PROP INTERFACES
// ============================================================================

// ============================================================================
// UTILITY TYPES
// ============================================================================

export type ExamStatusType = keyof typeof ExamStatus;
export type ExamTypeType = keyof typeof ExamType;
export type QuestionTypeType = keyof typeof QuestionType;
export type SessionStatusType = keyof typeof SessionStatus;

// ============================================================================
// EXAM ASSIGNMENTS & FORMS
// ============================================================================

// Exam List Response (from app/admin/exams/page.tsx)
export interface ExamListResponse {
  exams: Exam[];
  total: number;
  page: number;
  page_size: number;
  has_more: boolean;
}

export interface QuestionAnalytics {
  question_id: number;
  question_text: string;
  correct_count: number;
  incorrect_count: number;
  skip_count: number;
  average_time_seconds: number;
}

// Question Form Data (from app/admin/exams/create/page.tsx & exam-question-form.tsx)
export interface QuestionCreateFormData {
  question_text: string;
  question_type: string;
  points: number;
  is_required: boolean;
  options?: string[];
  correct_answer?: string | number;
  correct_answers?: (string | number)[];
  time_limit_seconds?: number;
  explanation?: string;
  difficulty_level?: string;
  tags?: string[];
}

// Exam Take Response (from app/exams/take/[examId]/page.tsx)
export interface ExamTakeResponse {
  session: SessionInfo;
  questions: Question[];
  answers: Answer[];
}

// Monitoring Event (from app/exams/results/[sessionId]/page.tsx)
export interface MonitoringEvent {
  id: number;
  event_type: string;
  event_data: any;
  timestamp: string;
}

// Exam Results (from app/exams/results/[sessionId]/page.tsx)

File: backend/app/schemas/exam.py
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field

# Import enums from models
from app.models.exam import ExamStatus, ExamType, QuestionType, SessionStatus

# Import TemplateQuestionSelection from question_bank schemas
from app.schemas.question_bank import TemplateQuestionSelection


class ExamBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    exam_type: ExamType = ExamType.CUSTOM
    time_limit_minutes: Optional[int] = Field(None, ge=1, le=480)  # Max 8 hours
    max_attempts: int = Field(1, ge=1, le=10)
    passing_score: Optional[float] = Field(None, ge=0, le=100)
    is_randomized: bool = False
    is_public: bool = False  # Public exams available to all companies
    allow_web_usage: bool = True
    monitor_web_usage: bool = False
    require_face_verification: bool = False
    face_check_interval_minutes: int = Field(5, ge=1, le=60)
    show_results_immediately: bool = True
    show_correct_answers: bool = False
    show_score: bool = True
    instructions: Optional[str] = None


class ExamCreate(ExamBase):
    company_id: Optional[int] = None  # NULL = global exam (system admin only)


class ExamUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    status: Optional[ExamStatus] = None
    time_limit_minutes: Optional[int] = Field(None, ge=1, le=480)
    max_attempts: Optional[int] = Field(None, ge=1, le=10)
    passing_score: Optional[float] = Field(None, ge=0, le=100)
    is_randomized: Optional[bool] = None
    is_public: Optional[bool] = None
    allow_web_usage: Optional[bool] = None
    monitor_web_usage: Optional[bool] = None
    require_face_verification: Optional[bool] = None
    face_check_interval_minutes: Optional[int] = Field(None, ge=1, le=60)
    show_results_immediately: Optional[bool] = None
    show_correct_answers: Optional[bool] = None
    show_score: Optional[bool] = None
    instructions: Optional[str] = None


class ExamInfo(ExamBase):
    id: int
    company_id: Optional[int]  # NULL for global exams
    created_by: Optional[int]
    status: ExamStatus
    created_at: datetime
    updated_at: datetime

    # Additional computed fields
    total_questions: Optional[int] = None
    total_sessions: Optional[int] = None
    completed_sessions: Optional[int] = None
    average_score: Optional[float] = None

    model_config = ConfigDict(from_attributes=True)


class ExamQuestionBase(BaseModel):
    question_text: str = Field(..., min_length=1)
    question_type: QuestionType
    order_index: int = Field(0, ge=0)
    points: float = Field(1.0, ge=0)
    time_limit_seconds: Optional[int] = Field(None, ge=1)
    is_required: bool = True
    options: dict[str, str] | None = None  # {"A": "Option 1", "B": "Option 2"}
    correct_answers: list[str] | None = None  # ["A"] or ["A", "B"]
    max_length: Optional[int] = Field(None, ge=1, le=10000)
    min_length: Optional[int] = Field(None, ge=0)
    rating_scale: Optional[int] = Field(None, ge=2, le=10)
    explanation: Optional[str] = None
    tags: list[str] | None = None


class ExamQuestionCreate(ExamQuestionBase):
    exam_id: int


class ExamQuestionUpdate(BaseModel):
    question_text: Optional[str] = Field(None, min_length=1)
    question_type: Optional[QuestionType] = None
    order_index: Optional[int] = Field(None, ge=0)
    points: Optional[float] = Field(None, ge=0)
    time_limit_seconds: Optional[int] = Field(None, ge=1)
    is_required: Optional[bool] = None
    options: dict[str, str] | None = None
    correct_answers: list[str] | None = None
    max_length: Optional[int] = Field(None, ge=1, le=10000)
    min_length: Optional[int] = Field(None, ge=0)
    rating_scale: Optional[int] = Field(None, ge=2, le=10)
    explanation: Optional[str] = None
    tags: list[str] | None = None


class ExamQuestionInfo(ExamQuestionBase):
    id: int
    exam_id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ExamQuestionPublic(BaseModel):
    """Public version of question (for candidates taking exam)"""

    id: int
    question_text: str
    question_type: QuestionType
    order_index: int
    points: float
    time_limit_seconds: Optional[int]
    is_required: bool
    options: dict[str, str] | None  # No correct answers exposed
    max_length: Optional[int]
    min_length: Optional[int]
    rating_scale: Optional[int]

    model_config = ConfigDict(from_attributes=True)


class ExamSessionCreate(BaseModel):
    exam_id: int
    assignment_id: Optional[int] = None


class ExamSessionUpdate(BaseModel):
    status: Optional[SessionStatus] = None
    current_question_index: Optional[int] = Field(None, ge=0)
    time_remaining_seconds: Optional[int] = Field(None, ge=0)
    user_agent: Optional[str] = None
    ip_address: Optional[str] = None
    screen_resolution: Optional[str] = None


class ExamSessionInfo(BaseModel):
    id: int
    exam_id: int
    candidate_id: int
    assignment_id: Optional[int]
    status: SessionStatus
    attempt_number: int
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    expires_at: Optional[datetime]
    time_remaining_seconds: Optional[int]
    current_question_index: int
    total_questions: int
    questions_answered: int
    score: Optional[float]
    max_score: Optional[float]
    percentage: Optional[float]
    passed: Optional[bool]
    web_usage_detected: bool
    web_usage_count: int
    face_verification_failed: bool
    face_check_count: int
    created_at: datetime
    updated_at: datetime

    # Exam info for convenience
    exam_title: Optional[str] = None
    exam_type: Optional[ExamType] = None
    candidate_name: Optional[str] = None
    candidate_email: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class ExamAnswerSubmit(BaseModel):
    question_id: int
    answer_data: dict[str, Any] | None = None
    answer_text: Optional[str] = None
    selected_options: list[str] | None = None
    time_spent_seconds: Optional[int] = Field(None, ge=0)


class ExamAnswerInfo(BaseModel):
    id: int
    session_id: int
    question_id: int
    answer_data: dict[str, Any] | None
    answer_text: Optional[str]
    selected_options: list[str] | None
    is_correct: Optional[bool]
    points_earned: float
    points_possible: float
    time_spent_seconds: Optional[int]
    answered_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ExamAssignmentCreate(BaseModel):
    exam_id: int
    candidate_ids: list[int]  # Can assign to multiple candidates
    due_date: Optional[datetime] = None
    custom_time_limit_minutes: Optional[int] = Field(None, ge=1)
    custom_max_attempts: Optional[int] = Field(None, ge=1, le=10)
    custom_is_randomized: Optional[bool] = None


class ExamAssignmentUpdate(BaseModel):
    due_date: Optional[datetime] = None
    custom_time_limit_minutes: Optional[int] = Field(None, ge=1)
    custom_max_attempts: Optional[int] = Field(None, ge=1, le=10)
    custom_is_randomized: Optional[bool] = None
    is_active: Optional[bool] = None


class ExamAssignmentInfo(BaseModel):
    id: int
    exam_id: int
    candidate_id: int
    assigned_by: Optional[int]
    due_date: Optional[datetime]
    custom_time_limit_minutes: Optional[int]
    custom_max_attempts: Optional[int]
    custom_is_randomized: Optional[bool]
    is_active: bool
    completed: bool
    notification_sent: bool
    reminder_sent: bool
    created_at: datetime
    updated_at: datetime

    # Related info for convenience
    exam_title: Optional[str] = None
    exam_type: Optional[ExamType] = None
    candidate_name: Optional[str] = None
    candidate_email: Optional[str] = None
    assigner_name: Optional[str] = None
    sessions_count: Optional[int] = None
    latest_session: Optional[ExamSessionInfo] = None

    model_config = ConfigDict(from_attributes=True)


class ExamMonitoringEventCreate(BaseModel):
    session_id: int
    event_type: str
    event_data: dict[str, Any] | None = None
    severity: str = "info"


class ExamMonitoringEventInfo(BaseModel):
    id: int
    session_id: int
    event_type: str
    event_data: dict[str, Any] | None
    severity: str
    timestamp: datetime

    model_config = ConfigDict(from_attributes=True)


class ExamResultSummary(BaseModel):
    """Summary of exam results for display"""

    session: ExamSessionInfo
    answers: list[ExamAnswerInfo]
    questions: list[
        ExamQuestionInfo
    ] | None = None  # Include if showing correct answers
    monitoring_events: list[ExamMonitoringEventInfo] | None = None


class ExamStatistics(BaseModel):
    """Statistics for an exam"""

    exam_id: int
    total_assigned: int
    total_started: int
    total_completed: int
    completion_rate: float
    average_score: Optional[float]
    average_time_minutes: Optional[float]
    pass_rate: Optional[float]
    question_statistics: list[dict[str, Any]]  # Per-question stats


class ExamListResponse(BaseModel):
    exams: list[ExamInfo]
    total: int
    page: int
    page_size: int
    has_more: bool


class ExamSessionListResponse(BaseModel):
    sessions: list[ExamSessionInfo]
    total: int
    page: int
    page_size: int
    has_more: bool


class ExamTakeRequest(BaseModel):
    """Request to start taking an exam"""

    exam_id: int
    assignment_id: Optional[int] = None
    test_mode: bool = False
    user_agent: Optional[str] = None
    screen_resolution: Optional[str] = None


class ExamTakeResponse(BaseModel):
    """Response when starting exam"""

    session: ExamSessionInfo
    questions: list[ExamQuestionPublic]
    current_question: Optional[ExamQuestionPublic] = None
    time_remaining_seconds: Optional[int]
    can_navigate: bool = True


class FaceVerificationSubmit(BaseModel):
    """Face verification data submission"""

    session_id: int
    image_data: str  # Base64 encoded image
    timestamp: datetime
    verification_type: str = "periodic"  # "initial", "periodic", "suspicious"


class FaceVerificationResponse(BaseModel):
    """Face verification result"""

    verified: bool
    confidence_score: float
    message: str
    requires_human_review: bool = False


# Hybrid Exam Schemas


class HybridExamQuestionCreate(ExamQuestionBase):
    """Question for hybrid exam (without exam_id, will be set during creation)"""

    pass


class HybridExamCreate(BaseModel):
    """Schema for creating a hybrid exam with custom questions + template selections"""

    # Base exam data
    exam_data: ExamCreate

    # Custom questions created specifically for this exam
    custom_questions: list[HybridExamQuestionCreate] = Field(
        default_factory=list, description="Custom questions created for this exam"
    )

    # Template selections (randomly select from question banks)
    template_selections: list[TemplateQuestionSelection] = Field(
        default_factory=list,
        description="Question bank selections for random question picking",
    )

    # Validation
    def validate_has_questions(self) -> bool:
        """Ensure at least one question source is provided"""
        return len(self.custom_questions) > 0 or len(self.template_selections) > 0


class HybridExamResponse(BaseModel):
    """Response after creating a hybrid exam"""

    exam: ExamInfo
    total_questions: int
    custom_count: int
    template_count: int
    selection_rules: dict[str, Any]  # The question_selection_rules stored in exam
